feat(heater-shaker): adding LED basic functionality

pmoegenburg opened the pull request.

Commit:
added some initial firmware

Commit:
completed callback sequence and added gcode sequence for turning on individual LEDs

Commit:
updated systemwide for use in c++ and c files

Commit:
updated c array to std array, system_wide variables, and debug fixes

Commit:
removed implemented SysTick_Handler

Commit:
Merge branch 'edge' into RET-3-heater-shaker-LED-bringup

sfoster1 commented on stm32-modules/heater-shaker/firmware/system/system_hardware.c:
```diff
@@ -60,3 +87,209 @@ asm volatile (
+    
+  /*##-4- Configure the NVIC for I2C ########################################*/   
+  /* NVIC for I2Cx */
+  HAL_NVIC_SetPriority(I2Cx_ER_IRQn, 0, 1);
```
These are really high priority and probably won't play well with calling freertos functions from within this ISR. we should try setting them to around I think 10 or so? The MAX_LIBRARY_PRIORITY value from freertoscofnig.

sfoster1 commented on stm32-modules/heater-shaker/firmware/system/system_hardware.c:
```diff
@@ -60,3 +87,209 @@ asm volatile (
+
+//bool system_hardware_set_led(uint8_t aTxBuffer[SYSTEM_WIDE_TXBUFFERSIZE], I2C_Operations operation) {
+bool system_hardware_set_led(uint8_t* aTxBuffer, I2C_Operations operation) {
+  //loop thru bits
```
these probably shouldn't be checked in I think - maybe store them as notes elsewhere?

sfoster1 commented on stm32-modules/heater-shaker/firmware/system/system_hardware.c:
```diff
@@ -60,3 +87,209 @@ asm volatile (
+  //returns bool for transmit_start_success
+
+  uint16_t base_register = 0;
+  uint8_t UpdateBuffer[1] = {};
```
[nit] let's be consistent with variable names - we're a mix of `camelCase`, `PascalCase`, and `snake_case` right now. I tend to like `snake_case` for variable names but really the big thing is consistency, whatever it is.

sfoster1 commented on stm32-modules/heater-shaker/firmware/system/system_hardware.c:
```diff
@@ -60,3 +87,209 @@ asm volatile (
+  }
+  return (status == HAL_OK);
+
+/*HAL_StatusTypeDef HAL_I2C_Mem_Write_IT(I2C_HandleTypeDef *hi2c, uint16_t DevAddress, uint16_t MemAddress,
```
should probably remove

sfoster1 commented on stm32-modules/heater-shaker/firmware/system/system_hardware.c:
```diff
@@ -60,3 +87,209 @@ asm volatile (
+/**
+  * @brief  I2C error callbacks.
+  * @param  I2cHandle: I2C handle
+  * @note   This example shows a simple way to report transfer error, and you can
```
update or remove

sfoster1 commented on stm32-modules/heater-shaker/firmware/system/system_hardware.c:
```diff
@@ -60,3 +87,209 @@ asm volatile (
+  */
+void HAL_I2C_ErrorCallback(I2C_HandleTypeDef *I2cHandle)
+{
+  /** 1- When Slave don't acknowledge it's address, Master restarts communication.
```
should probaly update or remove

sfoster1 commented on stm32-modules/heater-shaker/firmware/system/system_hardware.c:
```diff
@@ -60,3 +87,209 @@ asm volatile (
+  SYSTEM_HW_HANDLE->led_transmit_error_complete(&result);
+}
+
+//*********** was in example project, but think including is wrong and cause of build errors *************
```
yes, this stuff shoudln't be here, just the I2C stuff. let's remove it.

sfoster1 commented on stm32-modules/heater-shaker/firmware/system/system_stm32f3xx.c:
```diff
@@ -177,6 +177,31 @@ void SystemInit(void)
 
+/* Reset the RCC clock configuration to the default reset state ------------*/
+/* Set HSION bit */
+RCC->CR |= 0x00000001U;
```
Where'd this stuff come from? Do we really want it? Is it to support soft reset better? If so, we should note what it's needed for.

sfoster1 commented on stm32-modules/heater-shaker/include/heater-shaker/gcodes.hpp:
```diff
@@ -676,6 +675,68 @@ struct SetSerialNumber {
     }
 };
 
+struct SetLED { //make sure M998 isn't used. Just use this to turn them on, cycle power to reset
```
bit of a nit, but could this be `M998.D`? I think this is more of a debug gcode because we probably won't let the higher levels set things this directly - it'll be built into internal state changes, or more generic commands with semantics like "identify".

sfoster1 commented on stm32-modules/heater-shaker/include/heater-shaker/gcodes.hpp:
```diff
@@ -676,6 +675,68 @@ struct SetSerialNumber {
+    template <typename InputIt, typename InputLimit>
+    requires std::forward_iterator<InputIt> &&
+        std::sized_sentinel_for<InputLimit, InputIt>
+    static auto write_response_into(InputIt buf, InputLimit limit, bool success) -> InputIt {
```
do we need a special handler here? if we reuse the acknowledgement with the error code, we can either say just `M998.D OK\n` if it succeeded or return the error code if it fails.

sfoster1 commented on stm32-modules/heater-shaker/include/heater-shaker/gcodes.hpp:
```diff
@@ -676,6 +675,68 @@ struct SetSerialNumber {
+            return std::make_pair(ParseResult(), input);
+        }
+        std::array<uint8_t, SYSTEM_WIDE_TXBUFFERSIZE> UpdateBuffer = {};
+        uint8_t index = static_cast<uint8_t>(value_res.first.value());
```
definitely need to bounds check this index and return a failure if you specify e.g. `255`.

sfoster1 commented on stm32-modules/heater-shaker/include/heater-shaker/system_task.hpp:
```diff
@@ -173,6 +174,37 @@ class SystemTask {
+            messages::HostCommsMessage(response)));
+
+        //create a new LEDTransmitCompleteResponse Host_Comms_Task message w .success and .error vars
+        //what kind of gcode output does Ack message w/o .id create? Need to print success and error vars
```
We should probably defer sending an ok back until the transmit is successful, i.e. make the logic  in `StartSetLEDMessage`'s visitor look like this, where we send an ack-with-error on failure and nothing on success, and then have this function be the one that sends the success message.

sfoster1 replied:
out of date comment?

sfoster1 reviewed (changes requested):
The functionality looks pretty good but there's some stuff I think we should change:
- Let's use the already existing ack-with-error for responses rather than having the response formatter take a bool
- Let's defer success responses to completion of the transmit rather than queuing the transmit
- I think the interrupt priorities are too high
- Need to clean up a lot of out-of-date comments that are notes to self or from the examples
- Tests!

Commit:
fixed up initializations and added function to test turning all LEDs on at startup

Commit:
debug updates to test

Commit:
added clearing of the LED driver shutdown register, all LEDs tested positively

Commit:
commented current and pwm static values

Commit:
Merge branch 'edge' into RET-3-heater-shaker-LED-bringup

fsinapi committed:
feat(thermocycler-refresh): Initial thermocycler-refresh bringup (#214)

* Removed superfluous FindSTM32G4xx.cmake
* Moved `find_package` calls for OpenOCD and CrossGCC out of heater-shaker folder to avoid errors in thermocycler-refresh folder
* Set up thermocycler-refresh CMakeLists files to support compilation of stm32-cross target (have not tested with stm32-host). Also added a variable for the name of the project to reduce manual overhead of setting up new module with cmake files
* Added base set of files to set up compilation of an executable binary for STM32G4 thermocycler board.

* Squashed 'cmake/' changes from faa0270..8bcd4c4

8bcd4c4 feat:Add STM32G4 USB support (#20)
555cbde refactor(cross toolchain): add separate tool chains for arm M4 and M33 (#19)
4d151a4 fix(bsp): use the correct path to port for FreeRTOS in the L5 BSP (#17)
e36f470 feat(cmake-utils): Add L5 source (#16)
7d7f852 chore(cmake): Update changes for FreeRTOS (#15)
521a6c4 Merge pull request #14 from Opentrons/linux-symlink
8b92a64 fix: Specify the correct executable instead of creating a symling
e6ae846 fix: Add creation of symlink to clang executable on Ubuntu
5db868a chore: merge changes from ot3-firmware (#12)
8a05ce8 chore(FindClang): Use clang 12 (#10)
b08f4b5 chore: Add FindFakeIt.cmake (#11)
d59934c chore: add FindFreeRTOS.cmake
7846c04 chore: add FindFreeRTOS.cmake
8a08778 Merge pull request #8 from Opentrons/pull_svd_for_all_st_boards
7ec0fae Use commit hash instead
a640f8a fix(cmake): update the source for svd files to include support for G4 boards
ff83072 fix(BSPs): reverse driver compilation (#6)
b60e089 Merge pull request #5 from Opentrons/fix_bsp_package_version
8a8d6f5 fix(g4_bsp): use the correct version when grabbing the bsp for the g4 board
0b30b6f feat(boost): add FindBoost (#4)
a624372 fix(FindClang): always use downloaded clang-format (#3)
4796fa5 fix(FindClang): fix old osx builtin versioning (#2)
df9f86b Merge pull request #1 from Opentrons/add_G4_support
9775940 Rename BSP files and names
0fd49b5 feat(cmake-utils): add support to find STM32G4 BSP

git-subtree-dir: cmake
git-subtree-split: 8bcd4c444db5072d516fcf0f67519fb8de8de7dc

* Updated CMakePresets.json to reflect change from STM32GCCCrossToolchain.cmake to STM32CortexM4GCCCrossToolchain.cmake

* Added dummy thermocycler-refresh/simulator/main.cpp and fixed up CMakeLists to avoid name conflict with heater-shaker

* More cleanup to thermocycler-refresh cmake files for host compilation
* clang-tidy on thermocycler-refresh files

* Updates to thermocycler github workflow so it isn't just a copy of the heater-shaker
* Added a dummy test executable for thermocycler-refresh
* Addressed linting issue with a hal-copied header in thermocycler-refresh

* Formatted dummy test executable :)
* Updated bootloader vector for STM32G491
* Removed heater-shaker errors from thermocycler-refresh

Co-authored-by: pmoegenburg <pbm53@cornell.edu>

Commit:
LED I2C now using FreeRTOS take and give notification

fsinapi committed:
feat(thermocycler-refresh): usb bringup (#215)

* Added content to thermocycler-refresh README files
* Added files to enable USB communication on thermocycler-refresh board
* Added files for system task and comms task for thermocycler-refresh
* Added support for basic system gcodes: dfu and get/set system info
* Fixed issue with serial number flash api
* Added TIM7 as a system timebase and set up a bitbang pwm loop for heartbeat led

* Fixed up clock configuration for thermocycler refresh (boost APB clocks to max speed)
* Added ccmram to linker script (not sure why stmcube didn't include it...)

* Set up simulator & a couple basic tests for thermocycler-refresh
* Copied over main simulator files from heater-shaker
* Copied over tests for Host Comms Task and System Task

* Updated USB PID for thermocycler-refresh
* Removed ccmram section definition from linker script for thermocycler-refresh

Commit:
refactored, cleaned up, and added blinking functionality

Commit:
added LED message hooks

Commit:
white LED only on left and red LED only on right

fsinapi committed:
refactor(heater-shaker, thermocycler-refresh): Consolidate common code (#216)

* Restructured stm32-modules folder to use a common include/ folder. Looks like a lot of changes but it's mostly moving files around.

* Consolidated some of the header files that were duplicated across the heater-shaker and thermocycler-refresh

* Clang format
* Fixed incorrect version header include for version.cpp.in

* Moving more files around... now compiling a common-core library and added STM32G491 and STM32F303 common-module code directories.

* Added thermistor csv back to source control

* Added a github workflow for stm32 module common folder
* Moved some tests to the common folder
* Set up linting, formatting, and proper test targets for stm32 common folder

* Added a stm32-modules/common README and updated readmes for other stm32 modules to reflect new directory structure

* Removed simulator files from common & moved them back to their respective modules
* Removed extraneous include_directories lines to add `include/common/`
* Added simulator build to CI configuration
* Fixed a wrong include path for heater_thread

fsinapi committed:
feat(thermocycler-refresh): Create thermal tasks (#217)

* Added initial files for the Thermal Plate task
* Added a message for Thermal Plate Thermistor read completion

* Consolidated all `find_package(clang)` lines in Cmake files to reduce time to build stm32 modules
* Fixed issues with gdbinit file locations for Heater/Shaker and Thermocycler-Refresh (was looking in source directory instead of binary directory)

* Further fixes to CMakeLists
* Code formatting

* Added a Lid Heater Task for the thermocycler-refresh

* Added thermal plate & lid heater test policy files to let Thermocycler tests compile

* Added simulator files to compile in (mostly useless) Lid Heater and Thermal Plate tasks
* Code autoformatting

Commit:
cleanup and fixes

Commit:
cleanup

Commit:
eliminated RCC resetting at SystemInit

sfoster1 commented on stm32-modules/heater-shaker/firmware/system/stm32f3xx_hal_msp.c:
```diff
@@ -0,0 +1,97 @@
+/**
+  ******************************************************************************
+  * @file    I2C/I2C_TwoBoards_ComIT/Src/stm32f3xx_hal_msp.c
```
We should probably get some of this boilerplate out of here

sfoster1 commented on stm32-modules/heater-shaker/firmware/system/system_hardware.c:
```diff
@@ -60,3 +90,149 @@ asm volatile (
+  const TickType_t xMaxBlockTime = pdMS_TO_TICKS( 100 );
+  HAL_StatusTypeDef status;
+
+  configASSERT( xTaskToNotify == NULL );
```
are `configAsserts` the right way to handle this? should we instead be returning early and sending an error?

sfoster1 commented on stm32-modules/heater-shaker/firmware/system/system_hardware.c:
```diff
@@ -60,3 +90,149 @@ asm volatile (
+    (uint16_t)REGISTER_SIZE, (uint8_t*)ShutdownBuffer, (uint16_t)(sizeof(ShutdownBuffer)));
+  ulNotificationValue = ulTaskNotifyTake( pdTRUE, xMaxBlockTime );
+
+  if ((status == HAL_OK) && (ulNotificationValue == 1) && (I2cHandle.State == HAL_I2C_STATE_READY)) {
```
[nit] if we make this return early on error, there will be fewer nested `if`s and it'll be a bit easier to read

sfoster1 commented on stm32-modules/heater-shaker/firmware/system/system_hardware.c:
```diff
@@ -60,3 +90,149 @@ asm volatile (
+  HAL_StatusTypeDef status;
+
+  configASSERT( xTaskToNotify == NULL );
+  xTaskToNotify = xTaskGetCurrentTaskHandle();
```
can we wrap blocks like this in a function?

fsinapi committed:
feat(thermocycler-refresh): Add KS103J2G thermistor conversions (#218)

* Added thermistor table for the KS103J2 thermistor used on the thermocycler-refresh.

* Added support in thermistor_conversion for using different thermistors

* Fixed bug with lambda function for resistance lookup
* Automated formatting

* Added test cases for thermistor conversions

* Changed the thermistor_conversion::Conversion class to be templated. Now different projects can specify their own set of thermistor tables to generate, and everything to do with the table lookups is done through a Functor template. Concepts are neat!
* Also updated Cmake files to generated seperate thermistor tables for each project, with the correct subset of tables (NTC for the heater/shaker, KS for the thermocycler). `Common` generates both for unit test purposes.

* Forgot to format code

* Fixed an extra semicolon in generated thermistor tables
* Updated -lint targets to capture generated thermistor tables

* Updated heater_thread in simulator to use new converter

sfoster1 commented on stm32-modules/heater-shaker/firmware/system/system_hardware.h:
```diff
@@ -1,16 +1,65 @@
+#include <stdbool.h>
+#include "systemwide.h"
+
+#define I2C_ADDRESS        0x6C
```
these values are only used in `system_hardware.c`, can we have them at the beginning of that file? then we can keep this file for just things that need to be shared between `system_hardware.c` and `freertos_system_task.cpp`

sfoster1 commented on stm32-modules/heater-shaker/firmware/system/system_hardware.h:
```diff
@@ -1,16 +1,65 @@
+#define I2Cx                            I2C1
+#define RCC_PERIPHCLK_I2Cx              RCC_PERIPHCLK_I2C1
+#define RCC_I2CxCLKSOURCE_SYSCLK        RCC_I2C1CLKSOURCE_SYSCLK
+#define I2Cx_CLK_ENABLE()               __HAL_RCC_I2C1_CLK_ENABLE()
```
having these in this header, for instance, will probably cause compiler errors when it's included from c++

sfoster1 commented on stm32-modules/heater-shaker/include/heater-shaker/gcodes.hpp:
```diff
@@ -676,6 +675,116 @@ struct SetSerialNumber {
+
+struct IdentifyModuleStopLED {
+    /*
+    ** IdentifyModuleStopLED uses a random gcode, M999, adjacent to the identify
```
[nit] `M999` is an emergency stop handling gcode in a lot of other implementations. can we pick something else?

sfoster1 commented on stm32-modules/heater-shaker/tests/test_host_comms_task.cpp:
```diff
@@ -475,6 +475,88 @@ SCENARIO("message passing for ack-only gcodes from usb input") {
+                REQUIRE(set_led_message.aTxBuffer == aTxBufferTest);
+                REQUIRE(written_firstpass == tx_buf.begin());
+                REQUIRE(tasks->get_host_comms_queue().backing_deque.empty());
+                /*AND_WHEN("sending a good response back to the comms task") {
```
can we either uncomment this or remove it

sfoster1 reviewed (changes requested):
The functionality looks really good! I think the big things we _have_ to fix before merging are some commented out code; left over boilerplate that can be removed; and some defines in `system_hardware.h` that could be in `system_hardware.c` and would cause compiler warnings. There's also some nitpicks around structuring some of the code.

However, the structure as a whole looks great, and I really like the use of task notifications!

Commit:
tidied up some code and wrote most tests

Commit:
updated tests, simulator, and formatted

Commit:
formatted some tests

sfoster1 reviewed (approved):
Awesome stuff!

Commit:
Merge branch 'RET-3-heater-shaker-LED-bringup' into edge

Commit:
eliminated space to make systemide.h identical to initial merge candidate

Commit:
updated systemwide.hpp to systemwide.h

Commit:
updated TCR test files for use with systemwide.h

Commit:
updated to systemwide.h in TCR simulator files
